stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -241,25 +241,50 @@ BuiltinTypeInfo::BuiltinTypeInfo(TypeRefBuilder &builder,
               builder.readTypeRef(descriptor, descriptor->TypeName)))
 {}
 
-bool
-BuiltinTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
-                                          remote::RemoteAddress address,
-                                          int *extraInhabitantIndex) const {
-    if (getNumExtraInhabitants() == 0) {
-      *extraInhabitantIndex = -1;
+bool BuiltinTypeInfo::readExtraInhabitantIndex(
+    remote::MemoryReader &reader, remote::RemoteAddress address,
+    int *extraInhabitantIndex) const {
+  if (getNumExtraInhabitants() == 0) {
+    *extraInhabitantIndex = -1;
+    return true;
+  }
+  // If it has extra inhabitants, it could be an integer type with extra
+  // inhabitants (like a bool) or a pointer.
+  // Check if it's an integer first. The mangling of an integer type is
+  // type ::= 'Bi' NATURAL '_'
+  llvm::StringRef nameRef(Name);
+  if (nameRef.startswith("Bi") && nameRef.endswith("_")) {
+
+    // Drop the front "Bi" and "_" end, check that what we're left with is an
+    // integer.
+    llvm::StringRef natural = nameRef.drop_front(2).drop_back();
+    uint8_t unused;
+    if (!natural.getAsInteger(10, unused)) {
+      assert(getSize() < 8 && "Integer has more than 8 bytes!");
+      uint64_t rawValue;
+      if (!reader.readInteger(address, getSize(), &rawValue))
+        return false;
+
+      auto maxRawValue = (256 << (getSize() - 1)) - 1;
```

Comment:
This calculation will overflow if `getSize()` is greater than `4`. Not an issue if the above code bails out when `getSize()` is too big. Or this can do the math in a `uint64_t` to avoid overflow.